ДЗ на тему "Собственный IoC контейнер"

daria-grebneva opened the pull request.

daria-grebneva committed:
ДЗ на тему "Собственный IoC контейнер"

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,120 @@
+    private void processConfig(Class<?> configClass) {
+        checkConfigClass(configClass);
+
+        TreeMultimap<Integer, String> methodsMultiMap = getParsedAnnotations(configClass);
```
Избыточно. Достаточно списка `Method` отсортированного по `order` из аннотации висящей над этими методами. Тогда остальной код тоже упростится

daria-grebneva replied:
Поменяла использование TreeMultimap<Integer, String> на List<Method>

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,120 @@
+                    if (Objects.equals(v, method.getName())) {
+                        Object component = (method.getParameterCount() > 0)
+                                ? method.invoke(config, getComponentParameters(method))
+                                : method.invoke(config);
```
`invoke` и пустой массив может принять. Так что можно обойтись без проверки

daria-grebneva replied:
Поправлено

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    private static TreeMultimap<Integer, String> getParsedAnnotations(Class<?> configClass) {
+        TreeMultimap<Integer, String> methodsMultiMap = TreeMultimap.create();
+        Arrays.stream(configClass.getDeclaredMethods()).forEach(m -> {
+            Arrays.stream(m.getDeclaredAnnotations()).forEach(annotation -> {
+                AppComponent appComponent = ((AppComponent) annotation);
+                if (methodsMultiMap.containsValue(appComponent.name())) {
+                    throw new RuntimeException("Found components with the same names");
+                }
+                methodsMultiMap.put(appComponent.order(), appComponent.name());
+            });
+        });
+        return methodsMultiMap;
+    }
```
Не совсем
```
        Arrays.stream(configClass.getDeclaredMethods()).
                filter(m -> m.isAnnotationPresent(AppComponent.class))
                .sorted(Comparator.comparingInt(m -> m.getAnnotation(AppComponent.class).order()))
                .toList();
```
А дубликаты можно проверить в момент прохода по этому списку

daria-grebneva replied:
Спасибо, поправлено!

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,120 @@
+    @Override
+    public <C> C getAppComponent(Class<C> componentClass) {
+
+        List<Object> componentList = appComponents.stream().filter(c -> c.getClass().getSimpleName().contains(componentClass.getSimpleName())).collect(Collectors.toList());
```
На имена лучше не ориентироваться. Особенно на их вхождение друг в друга. Лучше по классам и с использованием `isAssignableFrom`

daria-grebneva replied:
Поправлено

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,120 @@
+    public <C> C getAppComponent(String componentName) {
+        Object component = null;
+
+        for (Map.Entry<String, Object> e : appComponentsByName.entrySet()) {
+            if (e.getKey().toLowerCase().contains(componentName.toLowerCase())) {
+                component = e.getValue();
+            }
+        }
```
Просто `appComponentsByName.get(componentName)`

daria-grebneva replied:
Мы же сохраняем в мапу в качестве ключа имя класса, например IOServiceStreams, и у нас должна быть возможность с данным ключом находить и по componentName "IOService", и "ioService" 
Поэтому иного варианта, как проверять на подстроку в ключе, я найти не могу. А вариант с get работать не будет. 
Возможно, я поняла задание неправильно, и в качестве ключа должно храниться не имя класса?

daria-grebneva committed:
Правки по комментариям

daria-grebneva committed:
Правки по комментариям

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,116 @@
+    private void processConfig(Class<?> configClass) {
+        checkConfigClass(configClass);
+
+        List<Method> parsedMethods = getParsedAnnotations(configClass);
```
Метод не возвращает распарсенные аннотации. Он возвращает аннотированные методы

daria-grebneva replied:
Поправила

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,116 @@
+
+        List<Method> parsedMethods = getParsedAnnotations(configClass);
+        Object config = initializeConfig(configClass);
+        Set<String> uniqueComponentNames = new HashSet<>();
```
Это не нужно. У нас есть ключи `appComponentsByName`. Можно проверять их наличие

daria-grebneva replied:
Поправила

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,116 @@
+            try {
+                Object component = m.invoke(config, getComponentParameters(m));
+                appComponents.add(component);
+                appComponentsByName.put(component.getClass().getSimpleName(), component);
```
Мапа называется компонентыПоИменам. Это значит, что туда кладем компоненты напротив их имени из аннотации `AppComponent`, а не имени класса

daria-grebneva replied:
Поправлено)

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,116 @@
+                appComponents.add(component);
+                appComponentsByName.put(component.getClass().getSimpleName(), component);
+            } catch (InvocationTargetException | IllegalAccessException e) {
+                LOGGER.error("Not possible to invoke method of this instance ", e);
```
Если в момент создания контейнера произошла ошибка, то приложение не должно запускаться

daria-grebneva replied:
Поправлено

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,116 @@
+
+    private Object[] getComponentParameters(Method method) {
+        Parameter[] parameters = method.getParameters();
+        List<Object> paramsList = new ArrayList<>();
+        for (Parameter parameter : parameters) {
+            Object param = this.getAppComponent(parameter.getType().getSimpleName());
+            paramsList.add(param);
+        }
+        return paramsList.toArray();
```
Зачем делать список, а потом из него делать массив, если можно сразу сделать массив?

daria-grebneva replied:
Логично, поправила

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,116 @@
+            o = configClass.getConstructor().newInstance();
+        } catch (InvocationTargetException | InstantiationException | IllegalAccessException |
+                 NoSuchMethodException e) {
+            LOGGER.error("Not possible to call instance of class ", e);
```
Не подавляем исключения

daria-grebneva replied:
Поправила

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,116 @@
+    public <C> C getAppComponent(String componentName) {
+        Object component = null;
+
+        for (Map.Entry<String, Object> e : appComponentsByName.entrySet()) {
+            if (e.getKey().toLowerCase().contains(componentName.toLowerCase())) {
+                component = e.getValue();
+            }
+        }
```
> Просто `appComponentsByName.get(componentName)`

> Возможно, я поняла задание неправильно, и в качестве ключа должно храниться не имя класса?

Да. Ответил выше

daria-grebneva replied:
Поправлено, спасибо!

daria-grebneva committed:
Правки по комментариям:
 - appComponentsByName теперь содержит "name" из аннотации компонента в качестве ключа
 - исключения теперь не гасятся

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,102 @@
+    private void processConfig(Class<?> configClass) {
+        checkConfigClass(configClass);
+
+        List<Method> parsedMethods = getAnnotatedMethods(configClass);
```
Метод поправили, переменную оставили

daria-grebneva replied:
Поменяла именование

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,102 @@
+        Object config = initializeConfig(configClass);
+
+        parsedMethods.forEach((m) -> {
+            String componentName = m.getAnnotation(AppComponent.class).name().toLowerCase();
```
Не нужно `toLowerCase`

daria-grebneva replied:
Поправила

stvort commented on hw13-ioc-container/homework/src/main/java/ru/otus/appcontainer/AppComponentsContainerImpl.java:
```diff
@@ -0,0 +1,102 @@
+        Parameter[] parameters = method.getParameters();
+
+        return Arrays.stream(parameters)
+                .map(p -> this.getAppComponent(p.getType().getSimpleName()))
```
У нас нет **никакой возможности** получать что-либо через `getAppComponent` по имени типа! У метода две перегрузки. Одна для получения компонента по его классу или классу его предков, другая по имени из аннотации, которое **никак не связано с именем типа**! То что сейчас имя из аннотации равно имени типа без учета регистра, это совпадение. Там могут быть любые имена

daria-grebneva replied:
Спасибо, я наконец-то поняла!

daria-grebneva committed:
Правки по комментариям
 - убрала костыли с именем метода, поменяла на name в аннотации
 - поправила именование переменной
 - убрала неиспользуемый логгер

daria-grebneva committed:
Merge branch 'main' into hw13-ioc-container
